src/lib.rs

```diff
@@ -1477,6 +2197,126 @@ impl<T> Drop for IntoIter<T> {
+    // then backshifts the array. During iteration the array is in an unsound state
+    // (big deinitialized hole in it), and this is very dangerous.
+    //
+    // Our first line of defense is the borrow checker: we have a mutable borrow, so nothing
+    // can access the ThinVec while we exist. As long as we make sure the ThinVec is in a valid
+    // state again before we release the borrow, everything should be A-OK! We do this cleanup
+    // in our Drop impl.
+    //
+    // Unfortunately, that's unsound, because mem::forget exists and The Leakpocalypse Is Real.
+    // So we can't actually guarantee our destructor runs before our borrow expires. Thankfully
+    // this isn't fatal: we can just set the ThinVec's len to 0 at the start, so if anyone
+    // leaks the Drain, we just leak everything the ThinVec contained out of spite! If they
+    // *don't* leak us then we can properly repair the len in our Drop impl. This is known
+    // as "leak amplification", and is the same approach std uses.
+    //
+    // But we can do slightly better than setting the len to 0! The drain breaks us up into
+    // these parts:
+    //
+    // ```text
+    //
+    // [A, B, C, D, E, F, G, H, _, _]
+    //  ____  __________  ____  ____
+    //   |         |        |     |
+    // prefix    drain     tail  spare-cap
+    // ```
+    //
+    // As the drain iterator is consumed from both ends (DoubleEnded!), we'll start to look
+    // like this:
+    //
+    // ```text
+    // [A, B, _, _, E, _, G, H, _, _]
+    //  ____  __________  ____  ____
+    //   |         |        |     |
+    // prefix    drain     tail   spare-cap
+    // ```
+    //
+    // Note that the prefix is always valid and untouched, as such we can set the len
+    // to the prefix when doing leak-amplification. As a bonus, we can use this value
+    // to remember where the drain range starts. At the end we'll look like this
+    // (we exhaust ourselves in our Drop impl):
+    //
+    // ```text
+    // [A, B, _, _, _, _, G, H, _, _]
+    // _____  __________  _____ ____
+    //   |         |        |     |
+    //  len      drain     tail  spare-cap
+    // ```
+    //
+    // And need to become this:
+    //
+    // ```text
+    // [A, B, G, H, _, _, _, _, _, _]
+    // ___________  ________________
+    //     |               |
+    //    len          spare-cap
+    // ```
+    //
+    // All this requires is moving the tail back to the prefix (stored in `len`)
+    // and setting `len` to `len + tail_len` to undo the leak amplification.
+    /// An iterator over the elements we're removing.
+    ///
+    /// As we go we'll be `read`ing out of the mutable refs yielded by this.
+    /// It's ok to use IterMut here because it promises to only take mutable
+    /// refs to the parts we haven't yielded yet.
+    ///
+    /// A downside of this (and the *mut below is that it makes this Invariant, when
```

Comment:
This sentence reads strangely. Are the parentheses in the wrong place? Also, "Inductive" doesn't need to be capitalized?